File: jgrpc/src/main/java/jgrpc/lib/symbol/Function.java
package jgrpc.lib.symbol;

import java.util.LinkedList;

public class Function extends Symbol {
    private Type retType;
    private LinkedList<Variable> args;

    public Function(String name, Type retType, Location location, LinkedList<Variable> args) {
        super(name, "global", location);
        this.retType = retType;
        this.args = args;
    }

    @Override
    public SymType getSymType() {
        return SymType.FUNC;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Function) {
            Function f = (Function) o;

            // Checks name, and return type
            if (!(getName().equals(f.getName()) && retType == f.retType)) {
                return false;
            }

            if (args.size() == f.args.size()) {
                for (int i = 0; i < args.size(); ++i) {
                    if (args.get(i).getType() != f.args.get(i).getType()) {
                        return false;
                    }
                }
            } else {
                return false;
            }

            return true;
        }

        return false;
    }
}

File: grpc/src/main/scala/grpc/lib/symbol/Function.scala
package grpc
package lib
package symbol

import grpc.lib.symbol.Type.Type

import scala.collection.mutable.ListBuffer

class Function(val name: String, private val retType: Type,
               val location: Location, private val args: ListBuffer[Variable])
  extends Symbol(name, "global", location) {

  def getArgs: ListBuffer[Variable] = args

  def getRetType: Type.Value = retType

  override def getSymType = SymType.FUNC
}
